### PR TITLE
Re-implement glob logic

### LINKED ISSUE #15: Confirm that all backends properly iterate over directories
Most cloud services don't have a notion of a "directory." We try to add this back in within our logic by doing our own path splitting.

We should make sure that the pathlib methods (`glob` and `iterdir`) and our methods do the same thing of:
 - Returning directories and/or files when iterating
 - Returning the root or current directory when being asked to iter or not
 - Sort order of the returned results (may not be possible/worth it since we use a generator)

### LINKED ISSUE #154: glob: difference between CloudPath and Paths
Hello,

I have a replicated filetree between my S3 compatible storage and a local filesystem.
However, there is a discrepency when I use the glob function.
The filesystem glob works fine, but I have to add `"**/*/"` at the beginning of the path for CloudPath's glob

```python
# Composition of the bucket (sorry for the long names)
# NB: there are duplicates with and without /, may be related to #148
>> list(cloud_path.glob("**/*"))
[S3Path('s3://my_bucket/MODULES/PROCESS/CLIPPING/'), 
S3Path('s3://my_bucket/MODULES/PROCESS/CLIPPING/DAX/'), 
S3Path('s3://my_bucket/MODULES/PROCESS/CLIPPING/DAX'), 
S3Path('s3://my_bucket/MODULES/PROCESS/CLIPPING/DAX/BY_PRODUCTS/'), 
S3Path('s3://my_bucket/MODULES/PROCESS/CLIPPING/DAX/BY_PRODUCTS'), 
S3Path('s3://my_bucket/MODULES/PROCESS/CLIPPING/DAX/BY_PRODUCTS/20191215T110441_S2_T30TXP_L2A_122756/'), 
S3Path('s3://my_bucket/MODULES/PROCESS/CLIPPING/DAX/BY_PRODUCTS/20191215T110441_S2_T30TXP_L2A_122756'), 
S3Path('s3://my_bucket/MODULES/PROCESS/CLIPPING/DAX/BY_PRODUCTS/20191215T110441_S2_T30TXP_L2A_122756/14_20191215T110441_S2_T30TXP_L2A_122756_CLIP.tif')]

# Not working ways of retrieving the file 14_20191215T110441_S2_T30TXP_L2A_122756_CLIP.tif
>> list(cloud_path.glob("DAX/BY_PRODUCTS/*/14_20191215T110441_S2_T30TXP_L2A_122756_CLIP.tif"))
[]
/14_20191215T110441_S2_T30TXP_L2A_122756_CLIP.tif')]
>> list(cloud_path.glob("**/DAX/BY_PRODUCTS/*/14_20191215T110441_S2_T30TXP_L2A_122756_CLIP.tif"))
[]
>> list(cloud_path.glob("*DAX/BY_PRODUCTS/*/14_20191215T110441_S2_T30TXP_L2A_122756_CLIP.tif"))
[]

# Correct way
>> list(cloud_path.glob("**/*DAX/BY_PRODUCTS/*/14_20191215T110441_S2_T30TXP_L2A_122756_CLIP.tif"))
[S3Path('s3://my_bucket/MODULES/PROCESS/CLIPPING/DAX/BY_PRODUCTS/20191215T110441_S2_T30TXP_L2A_122756]
```

Note: I am still on my Ceph compatible storage (see #148), I hope this is not related 😓

### LINKED ISSUE #186: Drop Python 3.6 support after 2021-12
Python 3.6 will stop receiving security updates after 2021-12:
https://www.python.org/dev/peps/pep-0494/#lifespan

We can remove it from our test matrix and our list of supported python versions in setup.py

We should replace it with Python 3.10 instead

### PR DESCRIPTION
**Pull request by pjbull**
`glob`, `rglob`, and `iterdir` did not work as their pathlib counterparts did.

In order to align our implementation with the `pathlib` versions, we adapted `CloudPaths` to work in the same way that the CPython implementation of `glob` does, [which is relatively complex in order to handle all of the glob scenarios](https://github.com/python/cpython/blob/3.10/Lib/pathlib.py#L385-L500). This is based on the `_Selector` class.

This change introduces two internal classes, `_CloudPathSelectable` and `_CloudPathSelectableAccessor` that wrap a `CloudPath` to expose the methods expected by `_Selector` and its children.

The change also changes the internal `Client._list_dir` API to return a tuple of the form (entry, is_dir), which could help address #181 and #176. Internal code where we use `iterdir` can be replaced with `_list_dir` if knowing if something is a file or directory is done within the method.

In order to close #15, this also makes sure that `iterdir` does not include `self`.

In order to make the tests pass, this also closes #186, which drops support for Python 3.6

Tests for all of these changes are included. The test suite is created to mirror the tests for `glob` and `rglob` in the CPython codebase. Note: `test_glob_many_open_files`, which emulates a CPython tests is pretty slow. We may consider not matching to the same depth and range to speed that up if it becomes frustrating.

Note: This does leverage some internals of CPython to work correctly. It's possible that implementations without the `_make_selector, _posix_flavour` private apis will not work. If this is a problem, we may need to re-implement or vendor `_Selector` and its child classes. It seems to at least work fine on `pypy` not sure about anything else:

```python
cloudpathlib on  glob-new-implementation [$?] 🐍 v3.7.12 via 🅒 cloudpathlib-pypy took 2s
❯ pypy
Python 3.7.12 | packaged by conda-forge | (44db2626, Nov 15 2021, 13:00:09)
[PyPy 7.3.7 with GCC Clang 11.1.0] on darwin

>>>> from cloudpathlib import CloudPath
>>>> root_dir = CloudPath("s3://drivendata-public-assets/")
>>>> list(root_dir.glob('**/*.txt'))
[S3Path('s3://drivendata-public-assets/nested_dir/test.dot/a_file.txt'), S3Path('s3://drivendata-public-assets/odsc-west-2019/DATA_DICTIONARY.txt')]
```

Closes #154
Closes #15
Closes #186

### CODE REVIEW

File: cloudpathlib/cloudpath.py [hunk 1]
```diff
@@ -305,28 +305,57 @@ def exists(self) -> bool:
...
-    def glob(self, pattern: str) -> Iterable["CloudPath"]:
-        # strip cloud prefix from pattern if it is included
-        if pattern.startswith(self.cloud_prefix):
-            pattern = pattern[len(self.cloud_prefix) :]
+    def _glob_checks(self, pattern):
+        if self.is_file():
+            raise NotImplementedError("Cannot run glob with a file as the root.")
+
+        if ".." in pattern:
+            raise NotImplementedError("Relative paths with '..' not supported in glob patterns.")
```
Comment (jayqi) [inline cloudpathlib/cloudpath.py:313]: I think we should make a new class in [`cloudpathlib.exceptions`](https://github.com/drivendataorg/cloudpathlib/blob/master/cloudpathlib/exceptions.py) like

```python
class CloudPathNotImplementedError(CloudPathException, NotImplementedError):
    pass
```

and use that.

File: cloudpathlib/cloudpath.py [hunk 2]
```diff
@@ -305,28 +305,57 @@ def exists(self) -> bool:
...
 
-        # identify if pattern is recursive or not
-        recursive = False
-        if pattern.startswith("**/"):
-            pattern = pattern.split("/", 1)[-1]
-            recursive = True
+    def glob(self, pattern):
+        """Iterate over this subtree and yield all existing files (of any
+        kind, including directories) matching the given relative pattern.
+        """
```
Comment (jayqi) [inline cloudpathlib/cloudpath.py:335]: We're copying docstrings from pathlib so I don't think we need this.

File: cloudpathlib/cloudpath.py [hunk 3]
```diff
@@ -305,28 +305,57 @@ def exists(self) -> bool:
...
+        pattern_parts = PurePosixPath(pattern).parts
+        selector = _make_selector(tuple(pattern_parts), _posix_flavour)
+
+        yield from self._glob(selector)
+
+    def rglob(self, pattern):
+        """Recursively yield all existing files (of any kind, including
+        directories) matching the given relative pattern, anywhere in
+        this subtree.
+        """
```
Comment (jayqi) [inline cloudpathlib/cloudpath.py:347]: Ditto on not needing the docstring.

File: cloudpathlib/cloudpath.py [hunk 4]
```diff
@@ -866,3 +892,81 @@ def _resolve(path: PurePosixPath) -> str:
...
+# Designed to be compatible when used by these selector implementations from pathlib:
+# https://github.com/python/cpython/blob/3.10/Lib/pathlib.py#L385-L500
+class _CloudPathSelectableAccessor:
+    def __init__(self, scandir_func):
+        self.scandir = scandir_func
+
+
+class _CloudPathSelectable:
+    def __init__(
+        self, relative_cloud_path: PurePosixPath, children: List, is_dir: bool, exists: bool = True
```
Comment (jayqi) [inline cloudpathlib/cloudpath.py:913]:
```suggestion
        self, relative_cloud_path: PurePosixPath, children: Dict[PurePosixPath, bool], is_dir: bool, exists: bool
```

- Fix type hint for `children`
- I think we don't need the default on `exists`. I think we always specify, and that's probably safer anyways.

File: cloudpathlib/cloudpath.py [hunk 5]
```diff
@@ -866,3 +892,81 @@ def _resolve(path: PurePosixPath) -> str:
...
+        return self._path.name
+
+    @staticmethod
+    @contextmanager
+    def scandir(root):
+        results = []
+        for c, is_dir in root._all_children.items():
+            if c.parent == root._path:
+                results.append(root._make_child_relpath(c.name))
+        yield results
```
Comment (jayqi) [inline cloudpathlib/cloudpath.py:946]:
```suggestion
        yield (
            root._make_child_relpath(c.name)
            for c, _ in root._all_children.items()
            if c.parent == root._path
        )
```

I think this should also work? Not sure if having a generator expression is clearer than the for loop, or if turning the list into an iterator is a meaningful optimization.

File: requirements-dev.txt [hunk 6]
```diff
@@ -16,7 +16,7 @@ pandas
 pillow
 psutil
 pydantic
-pytest
+pytest<7.0
```
Comment (jayqi) [inline requirements-dev.txt:19]: What happened here? Were there breaking changes?

File: setup.py [hunk 7]
```diff
@@ -59,5 +60,5 @@ def load_requirements(path: Path):
         "Source Code": "https://github.com/drivendataorg/cloudpathlib",
     },
     url="https://github.com/drivendataorg/cloudpathlib",
-    version="0.6.5",
+    version="0.6.6",
```
Comment (jayqi) [inline setup.py:63]: Should match release notes.

**Review by jayqi — CHANGES_REQUESTED** [review]

File: cloudpathlib/cloudpath.py [hunk 8]
```diff
@@ -305,28 +305,57 @@ def exists(self) -> bool:
     def fspath(self) -> str:
         return self.__fspath__()
 
-    def glob(self, pattern: str) -> Iterable["CloudPath"]:
-        # strip cloud prefix from pattern if it is included
-        if pattern.startswith(self.cloud_prefix):
-            pattern = pattern[len(self.cloud_prefix) :]
+    def _glob_checks(self, pattern):
+        if self.is_file():
+            raise NotImplementedError("Cannot run glob with a file as the root.")
```
[the remark this answers is not in the document]
Comment (pjbull) [reply cloudpathlib/cloudpath.py:310]: Agreed. Removed this check and the tests. Confirmed you can both read and glob `foo`:

image

File: requirements-dev.txt (hunk 6 above)
Comment (pjbull) [reply requirements-dev.txt:19]: Yeah, they just released 7.0 a couple days ago and our test suite breaks. There wasn't a simple/obvious fix to the error I was seeing. I'll file an issue to investigate.

**Comment by pjbull** [comment]
@jayqi review comments implemented, thanks!

**Review by jayqi — APPROVED** [review]
I'm still pretty amazed that we got this to work. Exciting stuff!

File: cloudpathlib/s3/s3client.py [hunk 9]
```diff
@@ -179,25 +179,30 @@ def _list_dir(self, cloud_path: S3Path, recursive=False) -> Iterable[S3Path]:
...
+                        True,
+                    )
 
                 # files in the directory
                 for result_key in result.get("Contents", []):
-                    yield self.CloudPath(f"s3://{cloud_path.bucket}/{result_key.get('Key')}")
+                    yield (
+                        self.CloudPath(f"s3://{cloud_path.bucket}/{result_key.get('Key')}"),
+                        False,
+                    )
```
Comment (jayqi) [inline cloudpathlib/s3/s3client.py:205]: @pjbull: Reviewing @GeorgeSabu's #206 (thanks!) made me realize a problem here. I think we need to both check whether `result_key` is size 0 and ends with trailing `/`, similar to what was done in #190.

**Review by jayqi — CHANGES_REQUESTED** [review]

File: cloudpathlib/s3/s3client.py (hunk 9 above)
Comment (pjbull) [reply cloudpathlib/s3/s3client.py:205]: Yeah, I wanted to keep fixing that issue separate, so I'm going to merge this and then rebase #206 onto it and test with some "fake" directories.

**Pull request merged**